src/types.rs

```diff
@@ -690,9 +696,13 @@ impl CK_ATTRIBUTE {
     }
   }
 
-  pub fn get_bool(&self) -> bool {
-    let data: CK_BBOOL = unsafe { mem::transmute_copy(&*self.pValue) };
-    CkFrom::from(data)
+  pub fn get_bool(&self) -> Result<bool, Error> {
+    if self.has_invalid_value() {
```

Comment:
can we add a private function which we can call something like `self.invalid_value()?;` please?

That would remove a lot of code duplication from all the changes here and below and basically just add a single line to the top of every `get_` function call. Take a look on how it is done for `Ctx` here: https://github.com/mheese/rust-pkcs11/blob/b47b97e38c9b895cd79d967aed4e06262617b51c/src/lib.rs#L270

Reply:
Good point, it would make it much more clear. I like having both the choice of a method checking if the value is valid with the method returning a `bool` and a method directly returning the `Result`. Both might be useful for a caller. Will change!

Reply:
to clarify: my idea is to have this as a private function so that in all the `get_` methods we can call this like `self.invalid_value()?;` instead of the if / else constructs which are the same everywhere and are just there to simply check if this is an invalid/unavailable value.